src/Features/Core/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.cs

```diff
@@ -22,111 +26,200 @@ public DiagnosticIncrementalAnalyzer(DiagnosticAnalyzerService owner, int correl
         }
 
         #region IIncrementalAnalyzer
-        public override Task AnalyzeDocumentAsync(Document document, SyntaxNode bodyOpt, CancellationToken cancellationToken)
+        public async override Task AnalyzeDocumentAsync(Document document, SyntaxNode body, CancellationToken cancellationToken)
         {
+            // ToDo: Should there be an exception handler here? Should there be checks for cancellation?
+
+            if (document.SupportsSemanticModel)
+            {
+                Project project = document.Project;
+                VersionStamp projectVersion = await project.GetDependentVersionAsync(cancellationToken).ConfigureAwait(false);
+
+                Compilation compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+                CompilationResult compilationResult = GetCompilationResult(compilation, project, projectVersion, cancellationToken);
+
+                if (!compilationResult.DocumentAnalyzed(document.Id) && !compilationResult.CompletionStarted)
+                {
+                    ImmutableArray<Diagnostic> diagnostics = await compilationResult.AnalyzeDocumentAsync(document, cancellationToken).ConfigureAwait(false);
+                    RaiseEvents(project, document, diagnostics);
+                }
+            }
+        }
+
+        public override Task AnalyzeSyntaxAsync(Document document, CancellationToken cancellationToken)
+        {
+            // Nothing happens here. The syntax tree analyzers get run as part of AnalyzeDocumentAsync.
             return SpecializedTasks.EmptyTask;
         }
 
         public override async Task AnalyzeProjectAsync(Project project, bool semanticsChanged, CancellationToken cancellationToken)
         {
-            var diagnostics = await GetDiagnosticsAsync(project.Solution, project.Id, null, cancellationToken).ConfigureAwait(false);
+            // ToDo: The intent of the version stamp below is to encode a difference if any document in the project or any reference has changed.
+            // Is that what the project version encodes, or does it just encode the configuration of the project?
+            // (A quick looks suggests that this is OK.)
+            VersionStamp projectVersion = await project.GetDependentVersionAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
actually, I think you can ignore my comment. since you are explicitly remove projectId when project is removed. you probably doesnt need to use ProjectState.